### PR TITLE
ARROW-11221: [Rust] DF Implement GROUP BY support for Float32/Float64

### PR DESCRIPTION
**Pull request by ovr**
Rust doesn't provide Eq, Hash for f32/f64 types inside stdlib, it's why I am using an external library called ordered-float which implements this traits. It's better to use external library instead of implementing own inside this repository.

### CODE REVIEW

**Comment by mqy** [comment]
No sure but I think it should be OK to use crate `ordered-float`:
- it has only one  rust file `https://github.com/reem/rust-ordered-float/blob/master/src/lib.rs` of about 800 lines.
- the crate `num_traits` we are using does not provide `Ord`, `Eq` traits.

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @ovr 

@alamb Do you think we should make this a feature due to the added dependency?

**Review by alamb — APPROVED** [review]
I agree with @mqy 's assessment that this is a relatively small new dependency (and it doesn't bring its own large depdneneyc stack https://github.com/reem/rust-ordered-float/blob/master/Cargo.toml) and so it should be fine.

File: rust/datafusion/src/physical_plan/hash_aggregate.rs [hunk 1]
```diff
@@ -776,6 +785,14 @@ pub(crate) fn create_group_by_values(
     for i in 0..group_by_keys.len() {
         let col = &group_by_keys[i];
         match col.data_type() {
+            DataType::Float32 => {
```
Comment (ovr) [inline rust/datafusion/src/physical_plan/hash_aggregate.rs:788]: I've updated PR, forget about this place. Testing it with real DB example.

Comment (alamb) [reply rust/datafusion/src/physical_plan/hash_aggregate.rs:788]: I suggest adding an end-to-end sql test in https://github.com/apache/arrow/blob/master/rust/datafusion/tests/sql.rs to make sure the plumbing is all hooked up correctly

Comment (ovr) [reply rust/datafusion/src/physical_plan/hash_aggregate.rs:788]: @alamb I've added tests, but I found a strange bug probably with count.

https://github.com/apache/arrow/pull/9175/commits/012c1ac0151878f43d60e3a7090f5b99947e3294#diff-4b06103cf2132b1ab297fbe8cd42622ecbe1109ea26df7d8b358fa36d739549cR357

Thanks

File: rust/datafusion/tests/sql.rs [hunk 2]
```diff
@@ -345,6 +345,50 @@ async fn csv_query_group_by_int_min_max() -> Result<()> {
...
+#[tokio::test]
+async fn csv_query_group_by_float32() -> Result<()> {
+    let mut ctx = ExecutionContext::new();
+    register_aggregate_floats_csv(&mut ctx)?;
+
+    let sql =
+        "SELECT COUNT(*) as cnt, c1 FROM aggregate_floats GROUP BY c1 ORDER BY cnt DESC";
+    let actual = execute(&mut ctx, sql).await;
+
+    // @todo Count OR GROUP BY has a bug?
```
Comment (ovr) [inline rust/datafusion/tests/sql.rs:357]: @andygrove @alamb @mqy 

Any ideas why COUNT not working with only one value?

Thanks

Comment (ovr) [reply rust/datafusion/tests/sql.rs]: Fixed, forget about CSV header...

**Review by alamb — APPROVED** [review]
This looks great -- thanks again @ovr

**Pull request closed without merging**